Optimize term count update during post status transition

pbearne opened the pull request:
The modification ensures that term counts are only updated when the post's status has truly changed. This reduces unnecessary operations, improving performance when there is no status change on a post.

Trac ticket: https://core.trac.wordpress.org/ticket/42522

Commit:
Optimize term count update during post status transition

The modification ensures that term counts are only updated when the post's status has truly changed. This reduces unnecessary operations, improving performance when there is no status change on a post.

peterwilsoncc reviewed (commented):
This change makes sense. 

My one concern is for custom taxonomies using the `update_count_callback` feature. Although I can not see how that would be a problem unless they're using the count column for a non-standard purpose.

Are you able to add some tests to ensure:

* Updating a post without changing the status doesn't attempt to update the count
  * I think checking the action `edited_term_taxonomy` does not fire will be suitable
  * Check that the count before and after is the same
* Updating a post to change the post status does update the count 
  * increases by one when published
  * decreases by one when unpublished
* Modifying the terms without changing the post status triggers a recount/count is as expected
* anything else you think I've missed.

Commit:
Add unit tests for _update_term_count_on_transition_post_status function

The commit adds various PHPUnit tests for the "_update_term_count_on_transition_post_status" function. The tests cover multiple scenarios including updates in term count when a post is published, moved to the trash, restored, permanently deleted, or removed from a term.

peterwilsoncc commented on tests/phpunit/tests/post/UpdateTermCountOnTransitionPostStatus.php:
```diff
@@ -0,0 +1,217 @@
+		wp_set_object_terms( $this->post_id, $this->term_id, $this->taxonomy );
+	}
+
+	/**
+	 * Tear down.
+	 */
+	public function tearDown(): void {
+		wp_delete_post( $this->post_id, true );
+		wp_delete_term( $this->term_id, $this->taxonomy );
+		unregister_post_type( $this->post_type );
+		unregister_taxonomy( $this->taxonomy );
+
+		parent::tearDown();
+	}
+
```
```suggestion
```

All of these are reset between tests by the test suite.

peterwilsoncc commented on tests/phpunit/tests/post/UpdateTermCountOnTransitionPostStatus.php:
```diff
@@ -0,0 +1,217 @@
+	 */
+	protected $taxonomy = 'category';
+
+	/**
+	 * Set up.
+	 */
+	public function setUp(): void {
+		parent::setUp();
+
+		register_post_type( $this->post_type, array( 'public' => true ) );
+		register_taxonomy( $this->taxonomy, $this->post_type, array( 'public' => true ) );
```
To allow for compatibility with various versions of phpunit, [WP uses `set_up()` from the Yoast PHPunit Polyfills.](https://github.com/Yoast/PHPUnit-Polyfills?tab=readme-ov-file#option-1-yoastphpunitpolyfillstestcasestestcase)

```suggestion
	/**
	 * Set up.
	 */
	public function set_up(): void {
		parent::set_up();

		register_post_type( $this->post_type, array( 'public' => true ) );
		register_taxonomy( $this->taxonomy, $this->post_type, array( 'public' => true ) );
	}
```

peterwilsoncc commented on tests/phpunit/tests/post/UpdateTermCountOnTransitionPostStatus.php:
```diff
@@ -0,0 +1,217 @@
+		register_post_type( $this->post_type, array( 'public' => true ) );
+		register_taxonomy( $this->taxonomy, $this->post_type, array( 'public' => true ) );
+
+		$this->post_id = self::factory()->post->create(
+			array(
+				'post_type'   => $this->post_type,
+				'post_status' => 'publish',
+			)
+		);
+
+		$this->term_id = self::factory()->term->create(
+			array(
+				'taxonomy' => $this->taxonomy,
+				'name'     => 'Test Category',
+			)
+		);
+
+		wp_set_object_terms( $this->post_id, $this->term_id, $this->taxonomy );
+	}
```
These can be shared fixtures so they don't need to be recreated be tween each test. Any changes within tests will be reset between test runs.

While updating, please check if you need to register the taxonomy and post type here too. I haven't done so.

```suggestion
	/**
	 * Create shared fixtures.
	 */
	public static function wpSetUpBeforeClass( WP_UnitTest_Factory $factory ) {
		self::$post_id = $factory->post->create(
			array(
				'post_type'   => self::$post_type,
				'post_status' => 'publish',
			)
		);

		self::$term_id = $factory->term->create(
			array(
				'taxonomy' => self::$taxonomy,
				'name'     => 'Test Category',
			)
		);

		wp_set_object_terms( self::$post_id, self::$term_id, self::$taxonomy );
	}
```

peterwilsoncc commented on tests/phpunit/tests/post/UpdateTermCountOnTransitionPostStatus.php:
```diff
@@ -0,0 +1,217 @@
+ */
+class Tests_Taxonomy_UpdateTermCountOnTransitionPostStatus extends WP_UnitTestCase {
+
+	/**
+	 * @var int Post ID.
+	 */
+	protected $post_id;
+
+	/**
+	 * @var int Term ID.
+	 */
+	protected $term_id;
```
```suggestion
	/**
	 * @var int Post ID.
	 */
	protected static $post_id;

	/**
	 * @var int Term ID.
	 */
	protected static $term_id;
```

peterwilsoncc reviewed (commented):
Thanks Paul, I've added a couple of suggestions fo the tests to avoid repeating work unnecessarily

pbearne committed:
Update tests/phpunit/tests/post/UpdateTermCountOnTransitionPostStatus.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/post/UpdateTermCountOnTransitionPostStatus.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/post/UpdateTermCountOnTransitionPostStatus.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

pbearne committed:
Update tests/phpunit/tests/post/UpdateTermCountOnTransitionPostStatus.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

Commit:
Refactor test variables to static in post status transitions

The commit substitutes instance variables for static ones in tests related to post status transitions. This refactor improves consistency in referencing these values and optimizes performance by invoking these values directly instead of through instance references.

peterwilsoncc commented on tests/phpunit/tests/post/UpdateTermCountOnTransitionPostStatus.php:
```diff
@@ -0,0 +1,210 @@
+	/**
+	 * Set up.
+	 */
+	public function set_up(): void {
```
```suggestion
	public function set_up() {
```

pbearne replied:
I thought I had to add the Void to make it match the YOAST polyfill

I will merge the change

peterwilsoncc reviewed (commented):
I can't figure out why the unrelated tests have started failing.

For the most part they look to be arrays returning in a different order but I can't see how this change would affect that.

pbearne committed:
Update tests/phpunit/tests/post/UpdateTermCountOnTransitionPostStatus.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

johnbillion committed:
Merge branch 'trunk' into 42522-Terms-are-uselessly-counted-when-saving-a-post

johnbillion committed:
Switch to an early return to minimise changed lines.

johnbillion committed:
This isn't needed.

johnbillion committed:
These tests don't touch the post status transition handling.

johnbillion committed:
Add a test to verify the terms aren't recounted when the status doesn't change.
